executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+        Box::new(
+            self.selected_tuple
+                .children()
+                .chain(self.multiplicity.iter()),
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusSend<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BusSend(tuple={}, interaction_id={})",
+            self.selected_tuple, self.interaction_id
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusReceive<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let multiplicity = match &self.multiplicity {
+            Some(m) => m.to_string(),
+            None => "None".to_string(),
+        };
+        write!(
+            f,
+            "BusReceive(tuple={}, multiplicity={}, interaction_id={})",
+            self.selected_tuple, multiplicity, self.interaction_id
+        )
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, derive_more::Display)]
+/// A constraint in the constraint system.
+/// It is similar to [powdr_ast::analyzed::Identity], but abstracts
+/// over (phantom) lookups, permutations, and bus interactions.
+/// Instead, they are represented as a (pair of) bus send and bus receive.
+pub enum Identity<T> {
+    Polynomial(PolynomialIdentity<T>),
+    Connect(ConnectIdentity<T>),
+    BusSend(BusSend<T>),
+    BusReceive(BusReceive<T>),
+}
+
+impl<T> Children<AlgebraicExpression<T>> for Identity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        match self {
+            Identity::Polynomial(i) => i.children_mut(),
+            Identity::Connect(i) => i.children_mut(),
+            Identity::BusSend(i) => i.children_mut(),
+            Identity::BusReceive(i) => i.children_mut(),
+        }
+    }
+
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        match self {
+            Identity::Polynomial(i) => i.children(),
+            Identity::Connect(i) => i.children(),
+            Identity::BusSend(i) => i.children(),
+            Identity::BusReceive(i) => i.children(),
+        }
+    }
+}
+
+impl<T> Identity<T> {
+    pub fn contains_next_ref(&self) -> bool {
+        // TODO: This does not check the definitions of intermediate polynomials!
+        self.children().any(|e| e.contains_next_ref())
+    }
+
+    pub fn id(&self) -> u64 {
+        match self {
+            Identity::Polynomial(i) => i.id,
+            Identity::Connect(i) => i.id,
+            Identity::BusSend(i) => i.id,
+            Identity::BusReceive(i) => i.id,
+        }
+    }
+}
+
+/// Converts a list of [powdr_ast::analyzed::Identity] into a list of [Identity].
+/// Polynomial and connect identities remain unchanged, phantom bus interactions
+/// are converted to either a bus send or bus receive, and permutations and lookups
+/// are converted to a pair of bus send and bus receive.
```

Comment:
Please add that it allocates IDs and thus can only be used on a full set of identities.

Reply:
Or maybe it's better to take an `Analyzed` so we can be sure? This also has the advantage that we can use this to properly detect constraints on multiplicities etc (and we also have intermediates).

Reply:
Good point!